server-spi/src/main/java/org/keycloak/storage/client/ClientLookupProvider.java

```diff
@@ -92,4 +94,14 @@
      * @return Stream of ClientModel or an empty stream if no client is found. Never returns {@code null}.
      */
     Stream<ClientModel> searchClientsByClientIdStream(RealmModel realm, String clientId, Integer firstResult, Integer maxResults);
+
+    /**
+     * Return all default scopes (if 'defaultScope' is true) or all optional scopes (if 'defaultScope' is false) linked with the client
+     *
+     * @param client Client
+     * @param defaultScopes if true default scopes, if false optional scopes, are returned
+     * @param filterByProtocol if true, then just client scopes of same protocol like current client will be returned
+     * @return map where key is the name of the clientScope, value is particular clientScope. Returns empty map if no scopes linked (never returns null).
+     */
+    Map<String, ClientScopeModel> getClientScopes(ClientModel client, boolean defaultScopes, boolean filterByProtocol);
```

Comment:
Does it make sense to return scopes without filtering by protocol? IOW - is `filterByProtocol` needed at all? (Indeed, `addClientScope` would need to ensure that a client scope is valid to the protocol of the client).

Reply:
When I search usages of original `ClientModel.getClientScopes(boolean, boolean)` there are some which use the option without filtering by protocol. e.g. here https://github.com/keycloak/keycloak/blob/master/server-spi-private/src/main/java/org/keycloak/models/utils/ModelToRepresentation.java#L595-L596

We can discuss this with @mposolda as he was the original author of client scopes, wdyt?

Reply:
Maybe "filterByProtocol" is not needed. But currently I am not 100% sure if removing "filterByProtocol" can have some side-effects during export/import for clients with changed protocol.

AFAIR there is one bad thing, that Keycloak currently allows to change the protocol of the client. We discussed with Stian and Pedro that it will be ideal to remove this possibility. Instead of changing protocol of the client from OIDC to SAML, you would need to remove the original client and add new client instead of directly changing protocol (as it doesn't makes sense to allow changing protocol as there is small common things to OIDC and SAML clients). JIRA is here https://issues.redhat.com/browse/KEYCLOAK-16559

Fix KEYCLOAK-16559 will simplify things, but since KEYCLOAK-16559 is still open, Keycloak currently allows that same client can contains both OIDC and SAML client scopes. If we remove this possibility and address KEYCLOAK-16559, then support for filtering by protocol won't be needed at all.

Reply:
I would agree that KEYCLOAK-16559 is a bug. The fact is though that there might be some scopes assigned to existing clients in existing installations that do not respect the client protocol. Hence the proposal is to perform the filtering by default in the provider (as if `filterByProtocol` was `true`) but do not expose this as a parameter to limit space for mistakes. Similarly, adding a scope that does not respect the protocol would be silently ignored in `addClientScopes` provider.

Would that work for you @mposolda, @vramik?

Reply:
It sounds ok for me, I'm just not sure at this moment about the usages of original method which are called with `filterByProtocol` set on `false`, we should double check it'd stay working as expected.

Reply:
As for the current usages, the `filterByProtocol == false` cases do not make sense. According to the code, they are used in these places:
- `ModelToRepresentation` and `RepresentationToModel` conversions - only stores / restores data but does not use it
- `ClientScopesClientRegistrationPolicy.beforeUpdate` - there is the check what client scopes to add but only those for the correct protocol would be added anyway if the proposal above was followed, effectively behaving like if `filterByProtocol == true`
- `KeycloakModelUtils.isClientScopeUsed` is only used by classes from that would be updated in this PR

All the rest is either tests or use `filterByProtocol == true` 

Hence by removing `filterByProtocol`, there is no functional change.

Reply:
I agree with remove `filterByProtocol` parameter. As long as existing tests are passing, we should be fine. From the long-term, it will be good to address KEYCLOAK-16559, which will make sure that client can't never have clientScopes of different protocols.